runtime/jitter/interval.go

```diff
@@ -0,0 +1,134 @@
+)
+
+// SetGlobalIntervalJitter sets the global interval jitter. It is safe to call
+// this method multiple times, but only the first call will have an effect.
+func SetGlobalIntervalJitter(p float64, rand *rand.Rand) {
+	globalIntervalJitterOnce.Do(func() {
+		globalIntervalJitter = Percent(p, rand)
+	})
+}
+
+// JitteredRequeueInterval returns a result with a requeue-after interval that has
+// been jittered. It will not modify the result if it is zero or is marked
+// to requeue immediately.
+//
+// To use this function, you must first initialize the global jitter with
+// SetGlobalIntervalJitter.
+func JitteredRequeueInterval(res ctrl.Result) ctrl.Result {
+	if !res.IsZero() && res.RequeueAfter > 0 {
+		res.RequeueAfter = globalIntervalJitter(res.RequeueAfter)
+	}
+	return res
+}
+
+// JitteredIntervalDuration returns a jittered duration based on the given
+// duration.
+//
+// To use this function, you must first initialize the global jitter with
+// SetGlobalIntervalJitter.
+func JitteredIntervalDuration(d time.Duration) time.Duration {
+	return globalIntervalJitter(d)
+}
+
+// IntervalOptions is used to configure the interval jitter for a controller
+// using command line flags. To use it, create an IntervalOptions and call
+// BindFlags, then call SetGlobalJitter with a rand.Rand (or nil to use the
+// default).
+//
+// Applying jitter to the interval duration can be useful to mitigate spikes in
+// memory and CPU usage caused by many resources being configured with the same
+// interval.
+//
+// When 1000 resources are configured to requeue every 5 minutes with a
+// concurrency setting of 50 and a process time of approximately 1 second per
+// resource.
+//
+// Without jitter, all 1000 resources will requeue every 5 minutes, resulting
+// in 50 resources requeueing simultaneously every second over a 20-second
+// window.
+//
+// However, when we apply +/-10% jitter to the interval duration, the requeueing
+// will be spread out over a 1-minute window. As a result, the number of
+// resources requeueing per second will vary between approximately 15 to 18.33.
+//
+// This smoother workload distribution can result in significant reductions in
+// the impact of CPU and memory spikes. This improvement in workload
+// distribution also translates into benefits for the Go garbage collector.
+// Notably, the garbage collector experiences reduced GC bursts and more
+// frequent collections, leading to improved overall performance.
+type IntervalOptions struct {
+	// Percentage of jitter to apply to interval durations. A value of 10
+	// will apply a jitter of +/-10% to the interval duration. It can not be negative,
+	// and must be less than 100.
+	Percentage uint8
+}
+
+// BindFlags will parse the given pflag.FlagSet and load the interval jitter
+// with the default value of 10%.
+func (o *IntervalOptions) BindFlags(fs *pflag.FlagSet) {
+	o.BindFlagsWithDefault(fs, -1)
```

Comment:
just curious, why not just pass 10 directly? i know it'll be set to 10 because its negative, but this way someone else reading the function would be instantly be sure that the default is 10%

Reply:
Last minute modifications and not spending sufficient time to think them through, really :-)